oxidation/libparsec_python/src/crypto.rs

```diff
@@ -52,3 +52,90 @@ impl HashDigest {
         }
     }
 }
+
+#[pyclass]
+#[derive(PartialEq, Eq)]
+pub(crate) struct SigningKey(parsec_api_crypto::SigningKey);
+
+#[pymethods]
+impl SigningKey {
+    #[new]
+    fn new(data: &[u8]) -> PyResult<Self> {
+        match parsec_api_crypto::SigningKey::try_from(data) {
+            Ok(h) => Ok(Self(h)),
+            Err(err) => Err(PyValueError::new_err(err)),
+        }
+    }
+
+    #[getter]
+    fn verify_key(&self) -> PyResult<VerifyKey> {
+        Ok(VerifyKey(self.0.verify_key()))
+    }
+
+    #[classmethod]
+    fn generate(_cls: &PyType) -> PyResult<SigningKey> {
+        Ok(SigningKey(parsec_api_crypto::SigningKey::generate()))
+    }
+
+    fn sign<'p>(&self, py: Python<'p>, data: &[u8]) -> PyResult<&'p PyBytes> {
+        Ok(PyBytes::new(py, self.0.sign(data).as_slice()))
+    }
+
+    fn __repr__(&self) -> PyResult<String> {
+        Ok(String::from("SigningKey()"))
+    }
+
+    fn __richcmp__(&self, py: Python, value: &SigningKey, op: CompareOp) -> PyObject {
+        match op {
+            CompareOp::Eq => (self == value).into_py(py),
+            CompareOp::Ne => (self != value).into_py(py),
+            _ => py.NotImplemented(),
+        }
+    }
+}
+
+#[pyclass]
+#[derive(PartialEq, Eq)]
+pub(crate) struct VerifyKey(parsec_api_crypto::VerifyKey);
+
+#[pymethods]
+impl VerifyKey {
+    #[new]
+    fn new(data: &[u8]) -> PyResult<Self> {
+        match parsec_api_crypto::VerifyKey::try_from(data) {
+            Ok(h) => Ok(Self(h)),
+            Err(err) => Err(PyValueError::new_err(err)),
+        }
+    }
+
+    fn verify<'p>(&self, py: Python<'p>, signed: &[u8]) -> PyResult<&'p PyBytes> {
+        match self.0.verify(signed) {
+            Ok(v) => Ok(PyBytes::new(py, &v)),
+            Err(err) => Err(PyValueError::new_err(err)),
```

Comment:
In Python implementation, a `nacl.exceptions.CryptoError` is raised

it is a bit messy to import `nacl` exception from here (given we don't explicitly depend on pynacl, but on on the other hand we cannot do `from parsec.crypto import CryptoError` given `parsec.crypto` depends on us...

Another solution would be to create a separate `parsec.crypto_exceptions` module (or replace `parsec/crypto.py` by a folder to have `parsec/crypto/impl.py` and `parsec/crypto/exceptions.py`)
@vxgmichel any opinion on this ?

Reply:
A way to do it is to use `use pyo3::import_exception;` and then `import_exception!(nacl.exceptions, CryptoError);`. It work perfectly for me @touilleMan